src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs

```diff
@@ -0,0 +1,249 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using System.Xml.Linq;
+using Hazelcast.Core;
+using Hazelcast.Networking;
+using Hazelcast.Testing;
+using Hazelcast.Testing.Logging;
+using Microsoft.Extensions.Logging;
+using NUnit.Framework;
+
+namespace Hazelcast.Tests.Networking
+{
+    public class RegressionWithRealNetworkTest : MultiMembersRemoteTestBase
+    {
+        private IDisposable HConsoleForTest()
+            => HConsole.Capture(options => options
+                .ClearAll()
+                .Configure<HConsoleLoggerProvider>().SetPrefix("LOG").SetMaxLevel()
+                .Configure().SetMinLevel().EnableTimeStamp(origin: DateTime.Now)
+                .Configure(this).SetMaxLevel().SetPrefix("TEST")
+            );
+
+        [TearDown]
+        public async Task TearDown()
+        {
+            //since tests are network related may need fresh members on cases.
+            await MembersOneTimeTearDown();
+        }
+
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task TestClientConnectionBeforeServerReady(bool smartRouting)
+        {
+            using var _ = HConsoleForTest();
+
+            var clientTask = CreateAndStartClientAsync(options =>
+            {
+                options.Networking.SmartRouting = smartRouting;
+                options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = int.MaxValue;
+                options.Networking.ReconnectMode = ReconnectMode.ReconnectAsync;
+                options.ClusterName = RcCluster.Id;
+            });
+
+            //Let the client tries to connect before cluster is ready
+            await Task.Delay(500);
+
+            var member = await AddMember();
+
+            var client = await clientTask;
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(ClientState.Connected, client.State);
+            }, 10_000, 500);
+        }
+
+        [Test]
+        [Timeout(20_000)]
+        [TestCase(true, "localhost", "127.0.0.1")]
+        [TestCase(true, "127.0.0.1", "localhost")]
+        [TestCase(true, "localhost", "localhost")]
+        [TestCase(false, "localhost", "127.0.0.1")]
+        [TestCase(false, "127.0.0.1", "localhost")]
+        [TestCase(false, "localhost", "localhost")]
+        public async Task TestConnectionCountAfterClientReconnect(bool smartRouting, string memberAddress, string clientAddress)
```

Comment:
What are we testing exactly with `memberAddress` vs `clientAddress`?

Reply:
This is ported from Java - so this is probably testing too many things including name resolution but... meh.